Mathlib/Algebra/Group/InjSurj.lean

```diff
@@ -183,7 +183,10 @@ admits an injective map that preserves `0` and `+` to an additive commutative mo
 protected def commMonoid [CommMonoid M₂] (f : M₁ → M₂) (hf : Injective f) (one : f 1 = 1)
     (mul : ∀ x y, f (x * y) = f x * f y) (npow : ∀ (x) (n : ℕ), f (x ^ n) = f x ^ n) :
     CommMonoid M₁ :=
-  { hf.commSemigroup f mul, hf.monoid f one mul npow with }
```

Comment:
Do you get the same performance benefit by switching to
```
  { hf.monoid f one mul npow, hf.commSemigroup f mul with }
```
I think the answer is probably yes.

Reply:
I wish someone would explain to me what is going on. I know that you and Matt have a good understanding of the phenomenon taking place here, but I don't get it yet and know sufficiently little about what's going on internally to be able to piece it all together.

Reply:
Trying in #11413

Reply:
Currently the structure instance elaborator tries to populate all the fields by one at a time, eta expanding any failed structure fields, and trying again. 

The problem with `with` was that it could never match on `toA` directly if you had an `A` in the sources itself or via some projection of the sources. The reason for this was very silly: `A` doesn't have a `self` field and we search through all fields in the structure for matching. 

This left the `toA` unfilled and resulted in the eta expansion in the next pass. 

The problem with the eta expansion is that unification defaults to eta expansion when two terms differ by a constructor application. So now where there might have been one quick check you are forced to check all the fields directly. If any of this contains data, then you are much worse off. 

So generally we want our terms to be as eta-reduced as possible still. Unification still takes them apart and looks at the pieces of they aren't. 

So we come to the issue here: Lean 4's preferred parent implementation means that if you can supply the instance of the preferred parent you want to do to avoid eta expansion in their flavor of inheritance. 

Another change that landed with `with` fix was that the structure instance elaborator should now always fill the field with the first occurrence in the sources reading left to right. If the preferred parent comes second, then you are in trouble because it's very likely Lean will not use it whole and the end product will be eta expanded much more than necessary. 

Here our preferred parent is `Monoid`. It should go first in the list of supplied sources. When it doesn't, Lean sees it has a `mul` from the `CommSemigroup` instance and uses that to build the `CommMonoid` instead of using `Monoid`. The result is that data, the multiplication, is exposed to unification each time this function is used to construct a `CommMonoid`